client/lib/analytics/page-view-tracker/index.jsx

```diff
@@ -102,9 +116,12 @@ const mapStateToProps = ( state ) => {
 		! currentSlug ||
 		( typeof currentSlug === 'number' && currentSlug === selectedSiteId ) ||
 		currentSlug === selectedSiteSlug;
+	const isSelectedSiteNotRequested =
+		currentSlug && ! selectedSiteId && ! state.sites.requesting[ currentSlug ];
 
 	return {
 		hasSelectedSiteLoaded,
+		isSelectedSiteNotRequested,
```

Comment:
as a passer-by I find the negation in this name hard to read or easy to get backwards. also, this appears to show `false` even when there is no selected site, which seems like a contradiction, also we're checking `state.sites.requesting` and being unfamiliar, does that connote that we're in the middle of a request or that at some point in time a site has been requested?

looking at this code I would have predicted a name more like `isRequestingCurrentSite` that drops the negations.
to capture the "at any point in time" nuance it would also possibly read clearer to follow the pattern on the line above this: `hasSelectedSiteBeenRequested` or `hasRequestedSelectedSite`.

in all of these cases I do find it less demanding cognitively to pass in the positive affirmation of the thing and then negate it in a conditional rather than to pass in the negated version and check that as a negative/absence.

```ts
if ( ! hasRequestedSelectedSite ) {
	return this.recordViewWithProperties();
}
```

thoughts?

Reply:
Your feedback about the naming makes sense. What I'm trying to capture here is the case when the URL contains a site in its segments, but data about the site is not requested (e.g. when you visit /pricing/:site in cloud while logged out).

This could be:
```js
const hasRequestedCurrentSite =
		currentSlug && ( selectedSiteId || state.sites.requesting[ currentSlug ] );
```
The `currentSlug` part checks that a site was passed in the URL. The other part checks whether the site data has been fetched already (`selectedSiteId`) or is being requested.

Reply:
are you saying that if we have a site selected that we know it's being requested? those seem like separate measures to me, an assumption that if we have a selected site then something out there in the framework must have requested it.

why not use `state.sites.requesting[ currentSlug ]` alone?

could you find a way to directly measure the logged-out state? we normally fetch the user information when logged in, and I'm guessing that when that's absent, and when we're not waiting on that fetch to resolve, then we know we're logged out. better yet, do we have data in state showing logged-in/logged-out?

it would absolutely be clearer to read `isLoggedOut` and know the intention was that. maybe even if we don't have a good way to solidify the logic behind calculating the value we can think about renaming the value _to the thing that we truly care about_.

Reply:
> why not use state.sites.requesting[ currentSlug ] alone?

When authenticated, it seems that the site data is always fetched before `PageViewTracker` is first rendered. `state.sites.requesting[ currentSlug ]` will be false, as will `hasRequestedCurrentSite` though the data was requested.

> are you saying that if we have a site selected that we know it's being requested? those seem like separate measures to me

My point is that if we have the id (as opposed to the slug, which is retrieved from the current URL), the data was requested and fetched successfully.

> could you find a way to directly measure the logged-out state?

I'm hesitant to go that way. The core of the issue is that if the site hasn't been requested, we still want to record the page view. I'd like to apply a fix (granted that it's clear enough) that stays close to that. Checking for the logged out state might dilute the intent.

Reply:
> if we have the id (as opposed to the slug, which is retrieved from the current URL)

we're not guaranteed to get a slug in Calypso. you can pass the id directly in the URL.

---

wow. hard problem. it would really be helpful to others (and probably your future self) to create a new specific name for this condition. `isUnauthenticatedViewer` or something. even if the logic is variable, that at least gives us a fixed point to "fix" and a specification against which to know if it's even broken.

maybe there's no better direct measure here, but it feels ad-hoc to infer from requesting state, especially given that we've already discovered there are important caveats to the data source.

> The core of the issue is that if the site hasn't been requested, we still want to record the page view.

do we have a good intuition for why we don't send it out if we haven't fetched the current site? I assume there must be a good reason (that I forget) otherwise wouldn't it be easier to remove that early-abort and _always_ send out the stat?

Reply:
> do we have a good intuition for why we don't send it out if we haven't fetched the current site?

My best guess is that at the time it was assumed that if a site was in context, the user would be authenticated. So this wasn't an issue before we introduced the Jetpack cloud pricing page, which is available whether you logged in or not.

> wouldn't it be easier to remove that early-abort and always send out the stat?

I'm afraid this would send an incomplete payload along the page view event, when the user is connected.

<hr>

I'll follow your advice and check whether the user is authenticated or not. It seems like that'll be easier to make sense of. Thanks for your input @dmsnell.